src/engine/app.rs

```diff
@@ -41,38 +44,78 @@ impl Application {
             self.timer.restart();
             self.advance_frame();
             self.timer.stop();
-            self.delta_time = self.timer.elapsed()
+            self.context.lock().unwrap().delta_time = self.timer.elapsed();
         }
 
         self.shutdown();
     }
 
     /// Sets up the application.
     fn initialize(&mut self) {
-        self.states.start();
+        self.states.start(self.context.lock().unwrap().deref_mut());
     }
 
     /// Advances the game world by one tick.
     fn advance_frame(&mut self) {
-        let engine_events = self.context.borrow_mut().poll_engine_events();
+        // let context = self.context.lock().unwrap().deref_mut();
+        let engine_events = self.context.lock().unwrap().poll_engine_events();
         for engine_event in engine_events {
-            self.context.borrow_mut().broadcaster.publish().with::<EngineEvent>(engine_event);
+            self.context.lock().unwrap().broadcaster.publish().with::<EngineEvent>(engine_event);
         }
-        let events = self.context.borrow_mut().broadcaster.poll();
-        self.states.handle_events(events);
-        while SteadyTime::now() - self.last_fixed_update > self.fixed_step {
-            self.states.fixed_update(self.fixed_step);
+        let events = self.context.lock().unwrap().broadcaster.poll();
+        self.states.handle_events(events, self.context.lock().unwrap().deref_mut());
+        let fixed_step = self.context.lock().unwrap().fixed_step.clone();
+        let last_fixed_update = self.context.lock().unwrap().last_fixed_update.clone();
+        while SteadyTime::now() - self.context.lock().unwrap().last_fixed_update > fixed_step {
```

Comment:
This doesn't seem to work, we get stuck here after some frames.

running cargo --example window with some added println!s.
```
app.rs states is running
window.rs Handling events!
app.rs advance frame after states.handle_events
app.rs states is running
app.rs advance frame after states.handle_events
app.rs states is running
app.rs advance frame after states.handle_events
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
app.rs in tick loop.
```